### ISSUE TITLE
ByteScheduler serious bug when cleaning up rendezvous

### ISSUE THREAD

**Issue by weidai0903** [bytedance/byteps#103]
In your tf.patch, following code were introduced:
int64 step_to_clean;
  {
    mutex_lock l(mu_);
    if (steps_to_clean_.size() > 1024) {
      step_to_clean = steps_to_clean_.front();
      steps_to_clean_.pop_front();
      steps_to_clean_.push_back(step_id);
    }
  }
  step_id = step_to_clean;

step_id is enqueued only when steps_to_clean_ has size bigger than 1024. Since steps_to_clean_ is initialized to have 0 elements, no step_id could really be enqueued. Therefore, this implementation is incorrect. Would you mind describing what this part of code basically trying to do here?

**Comment by ymjiang** [comment]
@yhpeng-git Please take a look at this.

**Comment by byronyi** [comment]
This should be fixed with an else branch which enqueues the step id. @weidai0903 mind to submit a PR?

**Comment by weidai0903** [comment]
> This should be fixed with an else branch which enqueues the step id. @weidai0903 mind to submit a PR?

@byronyi Even if you did that change which becomes
int64 step_to_clean;
  {
    mutex_lock l(mu_);
    if (steps_to_clean_.size() > 1024) {
      step_to_clean = steps_to_clean_.front();
      steps_to_clean_.pop_front();
    } else {
      steps_to_clean_.push_back(step_id);
    }
  }
  step_id = step_to_clean;
This still does not make much sense since cleaning each individual rendezvous should be triggered when dequeuing each step_id and NOT triggered when enqueuing. With that being said, cleaning each step does not stand alone with enqueuing and dequeuing as it currently does. Am I understanding this part of code correctly?

**Comment by lxl910915** [comment]
> This should be fixed with an else branch which enqueues the step id. @weidai0903 mind to submit a PR?

You'd better declare ByteScheduler is pseudo-code just for you paper no for production?

**Comment by yhpeng-git** [comment]
@lxl910915 Sure. We will clarify this in the README.

**Comment by byronyi** [comment]
@weidai0903 The logic we are trying to address here is as follows:

ByteScheduler opportunistically starts to transmit parameters for the (n+1)-th round if the corresponding gradients for the n-th round have been applied, even if the whole step (n) is not finished. However, this is not possible in vanilla distributed TensorFlow, as it cleans up the rendezvous at the end of each step, so it would be impossible to overlap the transmissions for (n+1)-th round parameters and n-th round gradients. This will defy the whole design of ByteScheduler.

This particular patch tries to address this issue by delaying the cleanup by a certain number of steps. I admit 1024 is a magic number, but an appropriate value will largely depend on the number of nodes of your training job, and also the size of available memory. For each step not being cleanup on time, the RendezvousManager will keep a table of Send/Recv tensors if they are not finished. In other words, if all Send/Recv tensors are finished (which they are, unless there is a bug in TF core) the extra memory overhead for not cleaning each step will only be an empty table, or a `std::unorder_map`, for each step. We think it is an acceptable overhead.

Hope the explanation help. Please let us know if you have a second thought. Thank you!

**Comment by bobzhuyb** [comment]
@lxl910915 It's demo code, not pseudo-code. We'll make sure you can reproduce the results in the paper in certain TF version and setups. However, you are right, it's not meant for production, so we are not merging it into the master branch.
